x-pack/plugins/event_log/server/saved_object_provider_registry.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { KibanaRequest, SavedObjectsClientContract } from 'src/core/server';
+
+import { fromNullable, getOrElse } from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
+
+export type SavedObjectGetter = (
+  ...params: Parameters<SavedObjectsClientContract['get']>
+) => Promise<unknown>;
+export type SavedObjectProvider = (request: KibanaRequest) => SavedObjectGetter;
+
+export class SavedObjectProviderRegistry {
+  private providers = new Map<string, SavedObjectProvider>();
+  private defaultProvider?: SavedObjectProvider;
+
+  constructor() {}
+
+  public registerDefaultProvider(provider: SavedObjectProvider) {
+    this.defaultProvider = provider;
+  }
+
+  public registerProvider(type: string, provider: SavedObjectProvider) {
+    if (this.providers.has(type)) {
+      throw new Error(
+        i18n.translate(
+          'xpack.eventLog.savedObjectProviderRegistry.registerProvider.duplicateProvider',
+          {
+            defaultMessage:
+              'The Event Log has already registered a Provider for the Save Object type "{type}".',
+            values: {
+              type,
+            },
+          }
+        )
+      );
+    }
+    this.providers.set(type, provider);
+  }
+
+  public getProvidersClient(request: KibanaRequest): SavedObjectGetter {
+    if (!this.defaultProvider) {
+      throw new Error(
+        i18n.translate(
+          'xpack.eventLog.savedObjectProviderRegistry.getProvidersClient.noDefaultProvider',
+          {
+            defaultMessage: 'The Event Log requires a default Provider.',
+          }
+        )
+      );
+    }
+
+    const scopedProviders = new Map<string, SavedObjectGetter>();
```

Comment:
I wasn't immediately sure what the context of `scoped` was here - then I realized this was the request-specific cache of getters - maybe `cachedProviders` would have been more obvious.  Of course, they are just cached across the lifetime of the request, so maybe that name is a little misleading as well.  Perhaps a comment is appropriate above this line, indicating the `scopedProviders` are a request-specific cache of saved object getters ...

Reply:
Sure - I was aligning it with the Elasticsearch & SO Scoped clients, as that's what it means there too. :)
Happy to add a comment.